src/context.cc

```diff
@@ -469,44 +469,62 @@ template <typename P> static uint32_t headerSize(const P &p) { return p ? p->siz
 FilterHeadersStatus ContextBase::onRequestHeaders(uint32_t headers, bool end_of_stream) {
   CHECK_HTTP2(on_request_headers_abi_01_, on_request_headers_abi_02_, FilterHeadersStatus::Continue,
               FilterHeadersStatus::StopIteration);
-  DeferAfterCallActions actions(this);
-  auto result = wasm_->on_request_headers_abi_01_
-                    ? wasm_->on_request_headers_abi_01_(this, id_, headers).u64_
-                    : wasm_
-                          ->on_request_headers_abi_02_(this, id_, headers,
-                                                       static_cast<uint32_t>(end_of_stream))
-                          .u64_;
-  if (result > static_cast<uint64_t>(FilterHeadersStatus::StopAllIterationAndWatermark))
+  uint64_t result;
+  {
+    DeferAfterCallActions actions(this);
+    result = wasm_->on_request_headers_abi_01_
+                 ? wasm_->on_request_headers_abi_01_(this, id_, headers).u64_
+                 : wasm_
+                       ->on_request_headers_abi_02_(this, id_, headers,
+                                                    static_cast<uint32_t>(end_of_stream))
+                       .u64_;
+  }
```

Comment:
Adding scope here will fire off all the deferred calls in the middle of this function. This significantly changes the flow control, and it's going to make bugs hard to reason about. I think you should revert those changes.